victor/src/layout/flow/mod.rs

```diff
@@ -135,23 +147,48 @@ fn layout_block_level_children<'a>(
     (child_fragments, absolutely_positioned_fragments, block_size)
 }
 
+fn collapse_adjoining_margins(first: Length, second: Length) -> Length {
+    match (first >= Length::zero(), second >= Length::zero()) {
+        (true, true) => first.max(second),
+        (false, false) => first.min(second),
+        _ => first + second,
+    }
+}
+
 fn place_block_level_fragment(
     fragment: &mut Fragment,
     current_block_direction_position: &mut Length,
+    ongoing_collapsed_margin: &mut Length,
 ) {
-    let (bpm, rect) = match fragment {
-        Fragment::Box(fragment) => (
-            fragment.padding.block_sum()
+    match fragment {
+        Fragment::Box(fragment) => {
+            if fragment.collapsible_margins {
```

Comment:
The logic here looks wrong.

There are three cases to consider. This fragment is:

* In flow, same-formatting-context block: its own margin are adjoining with that of its in-flow siblings, and can (depending on border and padding) be adjoining with that of its first and list child.
* Other in flow block-level: same as above for siblings, but children are never adjoining
* Out of flow: (This means float, since abspos fragments are stored separately and never end up in this function.) This fragment is ignored completely: a previous in-flow sibling and a next in-flow sibling can still be adjoining between them, so `ongoing_collapsed_margin` should not be reset to zero.